pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -601,6 +604,46 @@ protected void handleOutOfOrderEvent(Object currentComparisonValue, Object recor
     }
   }
 
+  /**
+   * When we have to process a new segment, if there are comparison value ties for the same primary-key within the
+   * segment, then for Partial Upsert tables we need to make sure that the record location map is updated only
+   * for the latest version of the record. This is specifically a concern for Partial Upsert tables because Realtime
```

Comment:
where the version come from?

Reply:
Version is computed using comparison column values. In case of ties, the latest record is determined to be of the latest version.